### PR TITLE
Change State for TIs without DagRun

### PR DESCRIPTION
**Pull request by kaxil**
---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).

### CODE REVIEW

File: airflow/jobs/scheduler_job.py [hunk 1]
```diff
@@ -592,7 +592,11 @@ def execute_callbacks(
                 elif isinstance(request, DagCallbackRequest):
                     self._execute_dag_callbacks(dagbag, request, session)
             except Exception:  # pylint: disable=broad-except
-                self.log.exception("Error executing callback for File: %s", request.full_filepath)
+                self.log.exception(
+                    "Error executing %s callback for file: %s",
+                    request.__class__.__name__,
+                    request.full_filepath
+                )
```
Comment (kaxil) [inline airflow/jobs/scheduler_job.py:599]: This is unrelated change, but address one of the comments from Tomek on the Scheduler HA PR

File: tests/jobs/test_scheduler_job.py [hunk 2]
```diff
@@ -1718,7 +1718,6 @@ def test_execute_task_instances_limit(self):
             ti.refresh_from_db()
             self.assertEqual(State.QUEUED, ti.state)
 
-    @pytest.mark.quarantined
```
Comment (kaxil) [inline tests/jobs/test_scheduler_job.py:1721]: Should I remove this change from here?

File: airflow/jobs/scheduler_job.py [hunk 3]
```diff
@@ -839,17 +843,17 @@ def _change_state_for_tis_without_dagrun(
...
+        :param dag_ids: TaskInstances associated with these DAG IDs
+            with states in the old_states will be examined
+        :type dag_ids: list[str]
         """
         tis_changed = 0
         query = session \
             .query(models.TaskInstance) \
             .outerjoin(models.DagRun, and_(
                 models.TaskInstance.dag_id == models.DagRun.dag_id,
                 models.TaskInstance.execution_date == models.DagRun.execution_date)) \
```
Comment (ashb) [inline airflow/jobs/scheduler_job.py:855]: We should change this to use the new relationship we've added

```suggestion
            .outerjoin(models.TaskInstance.dag_run) \
```

File: tests/jobs/test_scheduler_job.py (hunk 2 above)
Comment (ashb) [reply tests/jobs/test_scheduler_job.py:1721]: Probably not as part of this PR -- we don't know if its stable yet.

File: airflow/jobs/scheduler_job.py [hunk 4]
```diff
@@ -1483,10 +1487,23 @@ def _do_scheduling(self, session) -> int:
             session.expunge_all()
             # END: schedule TIs
 
-            # TODO[HA]: Do we need to call
-            # _change_state_for_tis_without_dagrun (2x) that we were before
-            # to tidy up manually tweaked TIs. Do we need to do it every
-            # time?
+            # TODO[HA]: Do we need to do it every time?
+            self._change_state_for_tis_without_dagrun(
+                dag_ids=self.dagbag.dag_ids,
```
Comment (ashb) [inline airflow/jobs/scheduler_job.py:1492]: At this point, this is basically ALL dags, isn't it? So we could change the logic to not take dag IDs at all.

File: airflow/jobs/scheduler_job.py [hunk 5]
```diff
@@ -1483,10 +1487,23 @@ def _do_scheduling(self, session) -> int:
...
+
+            self._change_state_for_tis_without_dagrun(
+                dag_ids=self.dagbag.dag_ids,
+                old_states=[State.QUEUED,
+                            State.SCHEDULED,
+                            State.UP_FOR_RESCHEDULE,
+                            State.SENSING],
+                new_state=State.NONE,
+                session=session
+            )
```
Comment (ashb) [inline airflow/jobs/scheduler_job.py:1506]:
```suggestion
            )
            
            session.commit()
```

File: airflow/jobs/scheduler_job.py [hunk 6]
```diff
@@ -872,7 +876,7 @@ def _change_state_for_tis_without_dagrun(
                     models.TaskInstance.execution_date ==
                     subq.c.execution_date) \
                 .update({models.TaskInstance.state: new_state}, synchronize_session=False)
```
Comment (ashb) [inline airflow/jobs/scheduler_job.py:878]: `ti.set_state` also sets end_time, and possibly duratiuon -- this was probably missing here! (Can be a separate PR)

Comment (kaxil) [reply airflow/jobs/scheduler_job.py:872]: Will do it in a separate PR

File: airflow/jobs/scheduler_job.py (hunk 5 above)
Comment (kaxil) [reply airflow/jobs/scheduler_job.py:1498]: Done

File: airflow/jobs/scheduler_job.py (hunk 4 above)
Comment (kaxil) [reply airflow/jobs/scheduler_job.py]: Updated: https://github.com/astronomer/airflow/pull/956/commits/6dd35bcc3986429ebed15b7287fc204a7d4fdcb0

File: airflow/jobs/scheduler_job.py (hunk 5 above)
Comment (kaxil) [reply airflow/jobs/scheduler_job.py:1498]: Updated: https://github.com/astronomer/airflow/pull/956/commits/6dd35bcc3986429ebed15b7287fc204a7d4fdcb0

File: airflow/jobs/scheduler_job.py (hunk 3 above)
Comment (kaxil) [reply airflow/jobs/scheduler_job.py]: Updated: https://github.com/astronomer/airflow/pull/956/commits/6dd35bcc3986429ebed15b7287fc204a7d4fdcb0

File: tests/jobs/test_scheduler_job.py (hunk 2 above)
Comment (kaxil) [reply tests/jobs/test_scheduler_job.py]: Updated: https://github.com/astronomer/airflow/pull/956/commits/6dd35bcc3986429ebed15b7287fc204a7d4fdcb0

File: airflow/jobs/scheduler_job.py [hunk 7]
```diff
@@ -1483,10 +1481,40 @@ def _do_scheduling(self, session) -> int:
...
+                #              (when NOWAIT isn't available)
+                db_err_code = getattr(e.orig, 'pgcode', None) or e.orig.args[0]
+
+                # We could test if e.orig is an instance of
+                # psycopg2.errors.LockNotAvailable/_mysql_exceptions.OperationalError, but that involves
+                # importing it. This doesn't
+                if db_err_code in ('55P03', 1205, 3572):
+                    self.log.debug("Lock held by another Scheduler")
+                else:
+                    raise
```
Comment (kaxil) [inline airflow/jobs/scheduler_job.py:1517]: Haven't added `session.rollback` here

Comment (kaxil) [reply airflow/jobs/scheduler_job.py:1517]: WDYT? We can add it before `raise` in the `else` block but might not need in `if` statement

Comment (ashb) [reply airflow/jobs/scheduler_job.py:1517]: I think you should encapsulate this in an `is_lock_not_available_error` function to start with -- it makes it much clearer.

We'll need a rollback in the `if` success block, otherwise the connection isn't usable for the next steps.

Comment (kaxil) [reply airflow/jobs/scheduler_job.py]: Updated

File: airflow/jobs/scheduler_job.py [hunk 8]
```diff
@@ -1526,6 +1533,24 @@ def _do_scheduling(self, session) -> int:
 
             return num_queued_tis
 
+    @staticmethod
+    def _is_lock_not_available_error(error: OperationalError):
```
Comment (ashb) [inline airflow/jobs/scheduler_job.py:1537]: In utils.sqlalchemy maybe?

Comment (kaxil) [reply airflow/jobs/scheduler_job.py]: done

File: tests/jobs/test_scheduler_job.py [hunk 9]
```diff
@@ -3641,35 +3640,35 @@ def test_execute_queries_count_no_harvested_dags(self, expected_query_count, dag
...
-            ([ 8,   8,   8,  8],  1,  1, "1d",  "None",  "no_structure"),  # noqa
-            ([ 8,   8,   8,  8],  1,  1, "1d",  "None",        "linear"),  # noqa
-            ([20,  12,  12, 12],  1,  1, "1d", "@once",  "no_structure"),  # noqa
-            ([20,  12,  12, 12],  1,  1, "1d", "@once",        "linear"),  # noqa
-            ([20,  22,  25, 28],  1,  1, "1d",   "30m",  "no_structure"),  # noqa
-            ([20,  22,  25, 28],  1,  1, "1d",   "30m",        "linear"),  # noqa
-            ([20,  22,  25, 28],  1,  1, "1d",   "30m",   "binary_tree"),  # noqa
-            ([20,  22,  25, 28],  1,  1, "1d",   "30m",          "star"),  # noqa
-            ([20,  22,  25, 28],  1,  1, "1d",   "30m",          "grid"),  # noqa
+            ([10,  10,  10, 10],  1,  1, "1d",  "None",  "no_structure"),  # noqa
```
Comment (ashb) [inline tests/jobs/test_scheduler_job.py:3643]: How come the query counts changed? Oh, this is writing to the S10nDag table(s) right?

**Review by ashb — APPROVED** [review]

File: tests/jobs/test_scheduler_job.py (hunk 9 above)
Comment (kaxil) [reply tests/jobs/test_scheduler_job.py:3643]: This test calls `job._do_scheduling` -> `_change_state_for_tis_without_dagrun` (2x) -> Updates Task Instance table.

So every time we call `job._do_scheduling` we will have these 2 queries
